PreservedObjectHandler.update_version_after_validation implemented

ndushay opened the pull request:
Implements `#update_version_after_validation` method in PreservedObjectHandler.

Also added FIXME notes in #confirm_version implementation methods.

closes #283 
closes #287

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -211,14 +247,16 @@ def with_active_record_transaction_and_rescue
   end
 
-  # expects @incoming_version to be numeric
-  def update_preserved_object(pres_obj, new_version)
```
note that I decided the `update_preserved_object` one liner method was stupid and better served with inline code.

jmartin-sul commented on spec/services/preserved_object_handler_update_version_spec.rb:
```diff
@@ -340,4 +334,370 @@
+            expect(pc.reload.last_audited).to be > orig_timestamp
+          end
+          it 'updates PreservedCopy last_checked_on_storage' do
+            orig_timestamp = pc.last_checked_on_storage
```
same (why no) `sleep` question

ndushay replied:
same answer:  last_checked_on_storage is aware of fractions of a second and needs no sleep.  Comments added to sleep lines.

jmartin-sul commented on app/services/preserved_object_handler.rb:
```diff
@@ -163,25 +195,29 @@ def create_db_objects(status, validated=false)
         results.concat(update_db_object(pres_copy))
         results << result_hash(ARG_VERSION_GREATER_THAN_DB_OBJECT, pres_object.class.name)
-        update_preserved_object(pres_object, incoming_version)
+        pres_object.current_version = incoming_version
```
what about also calling `update_pc_validation_timestamps(pres_copy) if validated` here?

jmartin-sul replied:
nah, that's fine, i missed that, and similar concerns here, but no suggestions for fixing them jump to mind

SaravShah commented on app/services/preserved_object_handler.rb:
```diff
@@ -163,25 +195,29 @@ def create_db_objects(status, validated=false)
         results << result_hash(ARG_VERSION_GREATER_THAN_DB_OBJECT, pres_copy.class.name)
-        update_preserved_copy(pres_copy, incoming_version, incoming_size)
-        results.concat(update_status(pres_copy, Status.ok))
+        update_preserved_copy_version_etc(pres_copy, incoming_version, incoming_size, validated)
```
what does _etc stand for?

ndushay replied:
"et cetera" because I couldn't think of a better name for the method.  Open to suggestions.

SaravShah replied:
updated_pc_obj
updated_pc_version
updated_pc_db_obj

i dunno just throwing names out there...

ndushay replied:
I was trying to say "this is updating a pc object version, along with some other fields"

SaravShah commented on spec/services/preserved_object_handler_update_version_spec.rb:
```diff
@@ -340,4 +334,370 @@
+          end
+        end
+
+        RSpec.shared_examples 'unexpected version with validation' do |incoming_version|
```
can this be moved to the `shared_examples_preserved_object_handler.rb` ?

ndushay replied:
It is only useful in this one context, so I left it here.  I know these tests are long;  I went with "shameless green".  Happy to entertain ideas for shortening them, etc.

SaravShah commented on app/services/preserved_object_handler.rb:
```diff
@@ -101,22 +101,54 @@ def confirm_version
       results << result_hash(INVALID_ARGUMENTS, errors.full_messages)
     else
-      Rails.logger.debug "update_version #{druid} called"
+      Rails.logger.debug "update_version_after_validation #{druid} called"
```
Do we need to know when the method was called in the logs?

ndushay replied:
yes.  It's a "debug" level message, so it will only be written if logging level is set to "debug"

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -101,22 +101,54 @@ def confirm_version
-        results.concat(update_online_version)
+        upd_results =
+          if moab_validation_errors.empty?
+            with_active_record_transaction_and_rescue do
```
good point.  I was applying copy pasta without thinking.  Changing.

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -101,22 +101,54 @@ def confirm_version
+      results << result_hash(INVALID_ARGUMENTS, errors.full_messages)
+    else
+      Rails.logger.debug "update_version #{druid} called"
+      upd_results = with_active_record_transaction_and_rescue do
```
i think we'll be fine without transaction wrapping here.  It's a lookup, not an update.  And there is no "else" fall through.  Removing transaction wrapping, captain.

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -163,25 +195,29 @@ def create_db_objects(status, validated=false)
-        # FIXME: TODO: should it update existence check timestamps/status?
+        results.concat(update_status(pres_copy, status)) if status
+        update_pc_validation_timestamps(pres_copy) if validated
+        results.concat(update_db_object(pres_copy)) if pres_copy.changed?
```
I took my spec from here:  https://github.com/sul-dlss/preservation_catalog/issues/287  which was from some conversations or other with Julian.  

Did I get it wrong?

jmartin-sul replied:
nope, my bad, i think i was just unclear on the expected behavior there

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -234,17 +272,19 @@ def version_comparison_results(db_object, version_symbol)
+  # FIXME: this needs to go away in favor of ? update_version_after_validation
+  #  it is only used by confirm_version, which should essentially call
+  #  update_version_after_validation if the incoming_version is higher than the db
+  # One big problem with this is the overwriting of the PC status without validating first
```
See #353

jmartin-sul replied:
thanks for filing!

ndushay commented on spec/services/preserved_object_handler_update_version_spec.rb:
```diff
@@ -340,4 +334,370 @@
+        end
+        it 'updates PreservedCopy last_checked_on_storage' do
+          orig_timestamp = pc.last_checked_on_storage
+          po_handler.update_version_after_validation
```
I only put in sleep where the tests failed without it.  I think it has to do with whether the time field has fractions of a second or not.  I can add comments to clarify.

ndushay commented on spec/services/preserved_object_handler_update_version_spec.rb:
```diff
@@ -340,4 +334,370 @@
+
+        it 'updates PreservedCopy last_audited' do
+          orig_timestamp = pc.last_audited
+          sleep 1
```
same answer about sleeping in tests.  :-D  (I think I deleted @jmartin-sul 's comment, by accident.)

ndushay commented on spec/services/preserved_object_handler_update_version_spec.rb:
```diff
@@ -340,4 +334,370 @@
+          expect(pc).not_to have_received(:touch)
+        end
+
+        context 'incoming version newer than catalog versions (both) (happy path)' do
```
good catch!

ndushay commented:
@jmartin-sul @SaravShah  I have added commits to address some of your concerns (and also rebased on master);  I responded to all other comments, I believe.
